### ISSUE TITLE
Implement gradient boosted trees with Numba

### ISSUE THREAD

**Issue by mrocklin** [dask/dask-ml#299]
It might be interesting to implement a Python-native gradient boosted tree solution.  LightGBM might be a good model to copy.

By being Python-native we would get a few benefits:

1.  We could use standard python data structures like Pandas dataframes rather than converting to XGBoost's DMatrix or similar
2.  Expert users could plug in their own objective functions, and so this would enable some creativity among the research crowd

@ogrisel might be able to provide some pointers about particular algorithms and pointers to codebases

cc @sklam @seibert @stuartarchibald

**Comment by ogrisel** [comment]
- A third benefit would be to directly tap into the dask API (probably the dynamic / eager / distributed scheduler) to implement distributed parallelism on a cluster of machines.

- A fourth benefit would be to leverage GPUs when available while keeping a single and readable code base.

The LightGBM algorithm is described in this paper https://papers.nips.cc/paper/6907-lightgbm-a-highly-efficient-gradient-boosting-decision-tree .

The [reference C++ implementation](https://github.com/Microsoft/LightGBM) is really clean and licensed under the MIT license so it should be fine to directly translate the meaty parts of the algorithm into `nopython` numba code: as long as we credit the original authors and keeps the same license for that part of the code base.

**Comment by ogrisel** [comment]
There are some benchmarks on standard medium to large scale datasets:

- single machine: https://github.com/Microsoft/LightGBM/blob/master/docs/Experiments.rst#comparison-experiment

- distributed (OpenMPI): https://github.com/Microsoft/LightGBM/blob/master/docs/Experiments.rst#parallel-experiment

**Comment by seibert** [comment]
This would be a very interesting application of Numba, and we'd welcome any feature requests that came out of it.  What level of effort do you think would this require? (ignoring the GPU case, which I think would be its own problem)

**Comment by mrocklin** [comment]
@ogrisel do you know of anyone else that would be interested in
collaborating on something like this?  My experience is that Numba devs are
really quite efficient when they're given numba code that mostly works, but
needs to work better.

**Comment by ogrisel** [comment]
Maybe @lancifollia would be interested :)

**Comment by ogrisel** [comment]
I will probably give it a try.

**Comment by lancifollia** [comment]
@ogrisel Thanks for mentioning me.
I'm interested in this project!

**Comment by ogrisel** [comment]
Great, let me push the code I wrote so far (not much yet).

**Comment by ogrisel** [comment]
@lancifollia @seibert I started an experimental repo here:

https://github.com/ogrisel/pygbm

I just started implementing two low-level primitives: 256-bits binning of numerical values and constructing histograms on such bins (work in progress).

I assume that the majority of the CPU time when fitting a decision tree will be spent constructing histograms.

@lancifollia please feel free to open a PR with a code dump of your pure python GBT + exact-method decision tree and so that we can start experimenting with some numba-ification of some of the performance critical functions of you exact algorithm.

I will go on working on the histogram for dense numerical data in the mean time.

**Comment by ogrisel** [comment]
BTW I don't plan to implement support for sparse data nor Gradient-based One-Side Sampling and Exclusive Feature Bunding from the LightGBM in the short term so as to keep the code as simple as possible for now.

**Comment by ogrisel** [comment]
I got a prototype working during my holidays :) It's very work in progress (only l2 loss, only dense numerical data, lacking tests, lacking doc...) but I found it quite easy to get something less than 2x slower than the highly optimized lightgbm for the Higgs boson dataset on my laptop.

https://github.com/ogrisel/pygbm/pull/11

@seibert To move forward it would be great to have a way to profile the body of jitted function calls. More details in https://github.com/numba/numba/issues/3254

**Comment by stsievert** [comment]
What updates are there to this issue? I've noticed that there is an upcoming SciPy 2019 talk on a Numbda implementation of gradient boosted trees, "Fast Gradient Boosting Decision Trees with PyGBM and Numba" by Nicolas Hug (who is working with @ogrisel).

**Comment by mrocklin** [comment]
Given the date I think that this came from a conversation that @ogrisel and I had at SciPy last year.  Yes, I think that this was implemented by Nicolas over the last year, and is now in Scikit-Learn proper.  

It might still be interesting to see if Dask could help this library scale, but that could be a separate issue.  I'll close this for now.  Thanks for pointing it out @stsievert .

**Comment by ogrisel** [comment]
Indeed there is numba implementation in PyGBM and a mirror Cython implementation in scikit-learn 0.21 / master.

The above implement the single node, multithreaded version of the lightgbm algorithm. However this algorithm is not directly amenable to distributed computation because the level where parallelism is used is would incur too much communication overhead if we were to do a naive distributed version of it.

To distribute this algorithm there is a variant that works with partitioned data as explained in this paper:

http://papers.nips.cc/paper/6380-a-communication-efficient-parallel-algorithm-for-decision-tree

This is probably what is implemented to run lightgbm with dask distributed in the https://github.com/dask/dask-lightgbm project.
